### ISSUE TITLE
Spike out push notifications triggered by some provider

### ISSUE THREAD

**Issue by oskarth** [status-im/status-react#1599]
Critical path for push notifications. Mock ability to trigger notifications from some provider.
Use https://github.com/zo0r/react-native-push-notification or https://github.com/evollu/react-native-fcm or whatever other library seems most mature, cross-platform, and flexible for multiple providers.

See https://docs.google.com/document/d/1OgjnY8ps8lVA4dIohwkfGK9HVt0nZxEWbuNdb7BX5-o/edit# for context

## Mock interfaces
onRegister: sends a token to register for push notifications. For now hardcode this to a specific provider, later on this will be dealt with differently.
onNotification: called when remote notification is received. For now this will simply open the app.

Also requires changing capabilities in Xcode and equivalent for Android in order to ask for permissions. Might require additional non-code changes (iTunes Connect etc).

## Goals
1. Ability to manually trigger a push notification.
2. Provide capabilities for normal push notifications so we have a starting point for more elaborate, encrypted and decentralized push notifications.
3. Clarify exact interface required from status-go in terms of tokens, payload structure, etc, as well as finding out any other unknowns we haven’t considered yet.

**Comment by jarradh** [comment]
We're using Firebase, so we'd be using https://github.com/invertase/react-native-firebase

**Comment by oskarth** [comment]
Push notifications require running on device, blocked by https://github.com/status-im/status-react/issues/1687

@jeluard pointed out push notifications work in the Android emulator, so I might switch the spike to focus on that path.

**Comment by oskarth** [comment]
Progress: able to register FCM token in latest Genymotion and then use this token to send notification via FB console. A start. This fulfills goal (1) albeit in a very ad hoc manner.

Notes:
- PN on Android emulator requires Google Play Service. This is possible with Open Gapps, but the latest binary in Genymotion 2.10 is broken. Instead we can use `open_gapps-x86-6.0-pico-20170825.zip` (from https://github.com/opengapps/opengapps) and drag and drop this on the emulator.
- Since PN aren't supported in iOS emulator, FB functions are compiled aways. This requires conditional loading of dependencies before WIP can be merged.

**Comment by oskarth** [comment]
Note for testing outside FB Console:

```
curl -X POST --header "Authorization: key=<SERVER KEY>" \ --Header "Content-Type: application/json" \ https://fcm.googleapis.com/fcm/send \ -d "{\"to\":\"<FCM TOKEN>\",\"notification\":{\"body\":\"test message here\"}}"
```

SERVER_KEY can be found in Firebase console for Status React App. FCM_TOKEN can currently be found with `adb logcat|grep token`.

**Comment by oskarth** [comment]
Progress: with workaround for https://github.com/status-im/status-react/issues/1687: I got a (slow) way of getting latest develop on release to work on iOS device. react-native-fcm appears to run fine and I can trigger permissions and get an FCMToken printed in debug logs.

This is good progress as it means we are closer to having two devices in dev for testing notifications. However, when I try to send a notification to this iOS device I get an Invalid APNs key. When I go to apple.developer.com to verify if this has been created correctly I find myself locked out due to some security questions. See https://github.com/status-im/status-react/issues/1658

**Comment by oskarth** [comment]
Managed to find old Apple session, generated APNS key and added to Firebase. Seems to work for iOS on device now, ish.

**Comment by oskarth** [comment]
As soon as #1749 is merged this issue should be considered to be closed, as the main purpose of the spike has been achieved. It has dragged out quite a while, largely due to build/tooling issues, which is not the purpose of a spike. There are still some questions left to answer but they are better treated as separate issues. We also want to avoid this becoming an umbrella issue for all the myriad little things that have to solve to solve the basic user story (https://docs.google.com/document/u/1/d/1OgjnY8ps8lVA4dIohwkfGK9HVt0nZxEWbuNdb7BX5-o/edit#) and integrate with status-go side of things.

**Comment by oskarth** [comment]
Issues with popup of FCM token not showing on device. Looking into it.

UPDATE: Fixed in PR, close this issue once release is out.

**Comment by oskarth** [comment]
Fixed in develop; will be in release 0.9.12.
